[WIP] Refactor instruction tracing

qwattash opened the pull request:
This replaces the instruction logging system we have in place for CHERI-MIPS.
The main goal is to make it more generic and consistent across architectures.
The current design removes the distinction between binary and text traces for targets and hides it behind a common interface. We log different parts of an instruction during execution as it was being done for cvtrace. Internally we build up a buffer for the current instruction information to log and we emit it at the end of the instruction translation. This unifies the interface and also permits to respect the -dfilter flags in a sensible way, so that if we are tracing a load to an address in the -dfilter address list we detect the match and emit the log for the whole instruction instead of just the memory-related event.

The current thinking is to move most instruction logging functionalities in separate common files:
- `include/qemu/log_instr.h` -> Arch-independent logging initialization and per-cpu state.
- `include/exec/log_instr.h` -> Interface for target helpers to the instruction logging
- `accel/tcg/log_instr.c` -> Main instruction logging implementation, including helpers

The implementation allows for easy addition of logging formats by defining a set of callbacks that a new format can define to emit various parts of a trace. The default format remains the "text" format, and can be switched with the `--cheri-trace-format` flag. The CPU_LOG_CVTRACE flag is dropped, now the CPU_LOG_INSTR flag is used to specify generically that instruction logging is active and there is a per-cpu flag to enable user-only logging which is used in conjunction with CPU_LOG_USER_ONLY.

qwattash committed:
Start implementing generic cheri instruction tracing interface.

arichardson commented on target/mips/op_helper.c:
```diff
@@ -2314,7 +2314,8 @@ enum {
 // Magic library function calls:
 void helper_magic_library_function(CPUMIPSState *env, target_ulong which)
 {
-    qemu_log_mask(CPU_LOG_INSTR, "--- Calling magic library function 0x" TARGET_FMT_lx "\n", which);
```
There is `qemu_log_mask_and_addr()`, maybe we should add a `qemu_log_instr_mask(env, mask, ...)` that expands to `if (unlikely(should_log_instr(env, mask)) { qemu_log(__va_args__)}`to avoid these ifs?

arichardson commented on target/mips/op_helper_log_instr.c:
```diff
@@ -665,21 +665,23 @@ static void dump_changed_regs(CPUMIPSState *env)
-        /* Print changed state: GPR, Cap. */
-        dump_changed_regs(env);
-    }
+        if (qemu_loglevel_mask(CPU_LOG_INSTR | CPU_LOG_CVTRACE)) {
```
This mask is equal to the toplevel if so can be dropped.

arichardson reviewed (commented):
Looks good, but maybe we should add a macro to avoid the extra `if`s?

qwattash committed:
Move instr logging start/stop helpers to generic implementation.

Move the instruction logging helpers from mips to generic interface.
Cleanup logging start and stop functions.
Fix generic instr logging helpers import in common helper headers.

qwattash committed:
Generalize instruction logging common instruction logging helpers.

This patch moves the logging helper generation to the generic translator loop,
it is no longer needed to modify the translation hook for the architecture to
commit the log buffer and log the next instruction PC.

Add a new translator hook to emit helpers to dump the per-architecture changed
CPU state. This is only called when the translator has instruction logging
enabled. It should be sufficient to check there as the TCG cache is flushed on
changes to the loglevel affecting this.

qwattash committed:
Move log info structure to generic CPUState.

qwattash committed:
Tentatively refactor core user mode logging.

User mode logging is now regulated by a per-cpu flag.
The function qemu_log_instr_enabled should be used to check for
instruction logging enabled state instead of the qemu_loglevel_mask
directly. This currently requires to have access to the CPUState
or CPUArchState wherever the check is performed, this is a problem
in riscv decoding as only the DisasContext is passed around.

qwattash committed:
Remove tranlator hook intended to dump changed state.

This is now done explicitly in exception handling and tr_insn if needed. The
reason behind this is that the hook has little value and in tr_insn we can pass
the current pc reliably to qemu_log_instr_mode_switch.

Consolidate the use of cached log_instr_enabled in DisasContextBase for checks.

qwattash committed:
Implment register logging interface.

General purpose registers are logged during translation. We ditch the
previous approach of scanning the register set for changed registers as
this causes confusion in traces where there might be instructions that
happen to write the same existing value in a register, resulting in the
update going detected and potentially picking up the value of an unrelated
register that changed as a side-effect of that instruction.
With the current approach we pick up ALL intendend register writes.

Port the MIPS implementation to the new logging interface. This logs
all GPR, CP0 and partially support capabilty register logging.
Tentatively port most of the mips implementation to the new GPR logging
interface.
Cop0 registers are reported as generic integer register changes in the
generic instruction logging interface.

arichardson commented on target/mips/cpu.c:
```diff
@@ -32,6 +32,133 @@
 #include "exec/exec-all.h"
 #include "qemu/error-report.h"
 
+#ifndef TARGET_MIPS64
```
If you can be bothered, using
```
const char mips_gp_regnames_flat[32][5] = {
    {"zero"}, {"at"}, {"v0"}, {"v1"}, {"a0"}, {"a1"},{"a2"}, {"a3"},
...
};
```
Reduces the binary size (and avoids relocations for PIE binaries)

arichardson commented on target/mips/translate.c:
```diff
@@ -10759,7 +10764,7 @@ static void gen_mftr(CPUMIPSState *env, DisasContext *ctx, int rt, int rd,
     }
     trace_mips_translate_tr("mftr", rt, u, sel, h);
-    gen_store_gpr(t0, rd);
+    gen_store_gpr(ctx, t0, rd);
```
I think for RISCV I avoided these diffs to upstream using this ugly hack:
`#define gen_store_gpr(val, rd) _get_store_gpr(ctx, val, rd)`
Not ideal but will reduce compilation failures after upstream merges.

qwattash replied:
I noticed that it does it. I was conflicted on whether to do the same and I avoided doing it for now on the basis that it is an ugly hack and that there are various other logging additions that may break merges in a similar way. I'll consider removing this diff if it makes it hard to merge upstream merges.

arichardson commented on accel/tcg/cputlb.c:
```diff
@@ -1751,7 +1751,7 @@ static inline uint64_t cpu_load_helper(CPUArchState *env, abi_ptr addr,
     ret = full_load(env, addr, oi, retaddr);
 
-#if defined(CONFIG_MIPS_LOG_INSTR)
+#if defined(CONFIG_CHERI_LOG_INSTR)
```
Ideally this should not be CHERI specific, maybe rename it to `CONFIG_TCG_LOG_INSTR` ?

qwattash replied:
I agree, I renamed it very early in the patch and should be changed. Will do.

arichardson commented on accel/tcg/log_instr.c:
```diff
@@ -0,0 +1,513 @@
+
+#ifdef CONFIG_CHERI_LOG_INSTR
+
+#ifdef DEBUG_INSTR_LOGGING
```
Should this automatically be on if CONFIG_DEBUG_TCG is on? DEBUG_TCG builds are already extremely slow, adding more assertions probably doesn't matter.

qwattash replied:
Good point.

arichardson commented on accel/tcg/log_instr.c:
```diff
@@ -0,0 +1,513 @@
+    /* Dump main instruction log */
+#if defined(TARGET_RISCV) && defined(CONFIG_RVFI_DII)
+    /*
+     * TODO(am2419): what to do with injected instructions?
```
I wouldn't worry about that now. If it's not I can fix it later. We don't really need instruction logging then other than for debugging failures.

arichardson commented on accel/tcg/log_instr.c:
```diff
@@ -0,0 +1,513 @@
+{
+    // TODO(am2419): how do we get the opcode from pc? this is a machine-dependant
+    // thing, maybe we can reuse disas or have an hook to determine the opcode length
+    // can we get it from next instruction pc?
```
You should be able to lookup the next pc from the saved TCG metadata. However, I'm not sure if that works when we reach the end of the translation block.

qwattash replied:
Thanks for confirming this, I will experiment a bit as I'm not sure yet.

arichardson commented on accel/tcg/log_instr.c:
```diff
@@ -0,0 +1,513 @@
+ *  instr logging-related and only used for this.
+ */
+/*
+void flush_tcg_on_log_instr_chage(void);
```
The only reason this ugly hack is needed is that log.c can't use tcg functions since it's also included in binaries such as qemu-img..

arichardson commented on accel/tcg/log_instr.c:
```diff
@@ -0,0 +1,513 @@
+    log_assert((mode & CPU_LOG_INSTR) ^ (mode & CPU_LOG_CVTRACE) &&
+        "Instruction logging must be enabled in either INSTR or CVTRACE mode");
+
+    if ((mode & CPU_LOG_USER_ONLY)) {
```
I wonder if we can remove the  CPU_LOG_USER_ONLY, and only allow toggling it from magic nops rather than the `-d` command line flag.

arichardson commented on accel/tcg/log_instr.c:
```diff
@@ -0,0 +1,513 @@
+        "Instruction logging must be enabled in either INSTR or CVTRACE mode");
+
+    if ((mode & CPU_LOG_USER_ONLY)) {
+        // TODO(am2419): looks like translator_ops::tb_in_user_mode(db, cpu) does
```
I added that and it will only work for RISCV+MIPS. Not sure if there is a better way of checking.

arichardson commented on accel/tcg/tcg-runtime.c:
```diff
@@ -169,26 +169,7 @@ void *HELPER(lookup_tb_ptr)(CPUArchState *env)
-        }
-    }
-}
+#if defined(CONFIG_CHERI_LOG_INSTR)
```
These should probably also move to the new logging source file?

qwattash replied:
These will probably be dropped when I finish the memory logging part.

arichardson reviewed (commented):
Haven't managed to look at all the changes yet, but generally looking good.

qwattash committed:
Rename instruction logging ifdefs.

Move CONFIG_CHERI_LOG_INSTR to CONFIG_TCG_LOG_INSTR.
Move CONFIG_INSTR_LOGGING ifdefs to CONFIG_DEBUG_TCG.

qwattash committed:
Implement instruction logging of memory accesses.

- Instrument the cpu_ld/st_mmuidx_ra interface (and sibling functions) to
  log memory accesses as non-capability memory accesses.
- Instrument the tcg_gen_qemu_ld/st interface to log memory accesses as
  non-capability memory accesses.
- Add extra functions cpu_ld/st_cap_data_ra interface to perform memory
  accesses for capability bytes without logging them.
- Add functions to log capability values as they are loaded/stored to memory.
  This uses compression routines in the implementation, tying to cheri 128
  compressed capabilities for now. The value associated to the memory access
  info is generally a duplicate of the one stored in the register access info
  for the same instruction. It is debatable if there are reasons to keep them
  both.

qwattash committed:
Cleanup checks for instruction logging enabled status.

Document and cleanup qemu_log_instr_enabled.

qwattash committed:
Remove or disable cvtrace everywhere.

This still leaves some stuff commented out, squash with future removal commit.

qwattash committed:
Implement new instruction logging exception interface.

Port the MIPS exception logging to the new interface.

qwattash committed:
Implement logging of ASID from targets.

Note that the CPU-ID will be emitted by the instruction logging backend as
it is available from the CPUState and is machine-independent.

qwattash committed:
Implement instruction logging output to legacy cvtrace.

qwattash committed:
Refactor the way trace format is handled.

- Add the notion of trace format within qemu_log_instr. The trace format can
  be set at startup using the --cheri-trace-format switch, as before.
- Ditch the CPU_LOG_CVTRACE flag.
- Fix user mode tracing enabling and mode switch.
- Ditch redundant user-mode switch checks and calls. Now the user mode switch
  is triggered in target-specific code. In MIPS this is done only on eret and
  when an interrupt is handled.

qwattash committed:
Remove all uses of CPU_LOG_CVTRACE.

This switches to use the qemu_log_instr format selection mechanism.
There is some code in MIPS magic nops that hexdump directy to the logfile,
  this breaks binary trace formats. Instead the dump is done to a string
  buffer, the string buffer is then appended to the log for the current
  instruction.

qwattash committed:
Cleanup all leftover functions.

Remove deprecated logging helpers.
Cleanup remaining uses of CPU_LOG_INSTR and qemu_loglevel_mask_* variants.

qwattash committed:
Implement -dfilter filtering for instruction logging.

Currently check the program counter and all the memory location accessed
  by an instruction.

qwattash committed:
Add missing helper to log updates to MIPS HI and LO registers.

qwattash committed:
Remove unused translator hook tb_in_user_mode.

This is removed because the translator does not check whether we are in user
mode directly anymore, the user mode logging switch done by each target on
exception/eret seems sufficient.

Rename cheri_get_asid to cpu_get_asid to keep consistent naming.

qwattash committed:
Remove debugging output from instruction logging.

arichardson commented on accel/tcg/cputlb.c:
```diff
@@ -1836,6 +1855,20 @@ uint64_t cpu_ldq_data_ra(CPUArchState *env, target_ulong ptr, uintptr_t retaddr)
+/*
+ * TODO(am2419): Ugly hack to avoid logging memory accesses that load capability
+ * components as normal memory accesses. The caller is responsible for logging.
+ */
```
I think logging the two memory reads is fine since it shows the PESBT and cursor values. Might be useful to debug compression issues instead of logging the two values separately afterwards.

That would also allow moving `log_instr_load_int` into cpu_load_helper.
If we want to avoid duplicate logging, I think having a MO_CAP is nicer long-term.

qwattash replied:
This was a point of contention for me. The reason I opted for this approach is that I want to distinguish explicitly that the value being loaded/stored are capability bytes.
I agree that ultimately we should treat capabilities as a new data type and handle both the tag and data within TCG ops, so having the two load/store separated breaks this idea of having `tcg_gen_cap_*` that operate on an hypothetical `target_cap`.
The solution implemented approaches what we would have with a `tcg_gen_cap_ld/st` interface.

Moreover, receiving the full capability register information gives us the capability fields and tag to store within the `log_meminfo_t` in the log info, without having to worry about the ordering of the write/reads.
If this was not the case, it would be hard to identify a set of writes as a capability+tag from the logging backend alone. This is relatively harmless for text tracing, but we need to reliably get the capability value if we want to support future uses of binary formats or other filtering that need to reason about the capability being written.
Note that this becomes an harder problem in presence of multiple writes, like aarch64 load/store pair.

arichardson commented on accel/tcg/cputlb.c:
```diff
@@ -2165,6 +2209,18 @@ void cpu_stq_data_ra(CPUArchState *env, target_ulong ptr,
+void cpu_stq_cap_data_ra(CPUArchState *env, target_ulong ptr,
+                         uint64_t val, uintptr_t retaddr)
+{
+    cpu_store_helper(env, ptr, val, cpu_mmu_index(env, false), retaddr, MO_TEQ);
```
Same here?

arichardson commented on accel/tcg/log_instr.c:
```diff
@@ -0,0 +1,902 @@
+    void (*emit_stop)(CPUArchState *env, target_ulong pc);
+    void (*emit_entry)(CPUArchState *env, cpu_log_instr_info_t *log);
+};
+typedef struct trace_fmt_hooks trace_fmt_hooks_t;
```
You could merge these typedefs with the struct delcaration? `typdef struct foo { ... } foo_t;`.

arichardson commented on accel/tcg/log_instr.c:
```diff
@@ -0,0 +1,902 @@
+
+            entry.entry_type = CTE_CAP;
+            entry.val2 = cpu_to_host64(metadata);
+            entry.val3 = cpu_to_host64(cap_get_cursor(cr));
```
cap_get_cursor returns a host endian value. If the cvtrace format is big endian you should use cpu_to_be64.

qwattash replied:
Right, I got it backwards.

arichardson commented on accel/tcg/log_instr.c:
```diff
@@ -0,0 +1,902 @@
+    entry.entry_type = CTE_NO_REG;
+    entry.thread = (uint8_t)env_cpu(env)->cpu_index;
+    entry.asid = (uint8_t)log->asid;
+    entry.pc = cpu_to_host64(log->pc);
```
same here.

arichardson commented on accel/tcg/log_instr.c:
```diff
@@ -0,0 +1,902 @@
+/*
+ * target-endiannes bswap helpers
+ */
+#ifdef TARGET_WORDS_BIGENDIAN
```
I don't think these are necessary.

arichardson commented on accel/tcg/translator.c:
```diff
@@ -127,11 +104,14 @@ void translator_loop(const TranslatorOps *ops, DisasContextBase *db,
+        /*
+         * TODO(am2419): could move the commit below tr_insn or would we lose something?
+         */
+        if (qemu_log_instr_enabled(cpu->env_ptr)) {
```
Can this use `db->log_instr_enabled` instead of calling the check function again? I believe anything that changes instruction logging state should break out of the translation block.
In that case I'd also be tempted to keep a local `const bool` variable since the compiler probably doesn't know that db->log_instr_enabled is constant.

qwattash replied:
It should definitely use the `db->log_instr_enabled`. I will also use a local `const bool`, nice catch.
I should also cleanup some of the TODOs and commented out stuff.

arichardson commented on util/log.c:
```diff
@@ -73,8 +73,8 @@ static void qemu_logfile_free(QemuLogFile *logfile)
-__attribute__((weak)) void flush_tcg_on_log_instr_chage(void) {
+__attribute__((weak)) void qemu_log_instr_flush_tcg(void);
+__attribute__((weak)) void qemu_log_instr_flush_tcg(void) {
     // Real implementation in translate-all.c
```
`// Real implementation in accel/tcg/log_instr.c since we can't use invoke tcg functions here`

arichardson commented on target/riscv/cpu.h:
```diff
@@ -586,4 +591,29 @@ riscv_cpu_get_tb_cpu_state(CPURISCVState *env, target_ulong *pc,
+}
+
+// TODO(am2419): deprecate and remove 
+// TODO(am2419) should probably rename as cpu_get_asid()
```
That TODO has been addressed.

I'm not sure we really need this function, but it was useful in the past to filter logs.

arichardson commented on target/mips/cpu.h:
```diff
@@ -1518,6 +1485,42 @@ static inline void mips_update_pc(CPUMIPSState *env, target_ulong pc_addr, bool
+/*
+ * Target-specific hook to fetch the cpu log state
+ */
+static inline cpu_log_instr_info_t *cpu_get_log_instr_state(CPUMIPSState *env)
```
Should this be part of the target-indepedent data instead?

qwattash replied:
I originally stored the `cpu_log_instr_info_t` in `CPUState` and access it in a target-independent way. I reverted back to having it target-specific because I needed to use target-specific types in the structure. Since I have no control over the naming or placement of the `cpu_log_instr_info_t` within `CPUArchState` I defined this.

qwattash replied:
On second thought, I think I can move it back to the `CPUState` and either have a `GArray` of log info or some other semi-opaque struct that gets allocated in `qemu_log_instr_init`. This would also align with the requirements of the ring-buffer idea.

arichardson reviewed (commented):
I haven't check that it actually works, but code generally looking good. Big improvement over the old implementation!

arichardson commented:
As a follow-up it might be nice to run in a mode where we have a ringbuffer of up to say 1000 instructions and dump that buffer to the logfile on certain events (e.g. CHERI trap). 
That would probably be a lot faster than doing a full trace while still having all the useful information.

qwattash commented:
The ring-buffer idea is really interesting! I think it goes into a different patch-set but I'd really like to have it.
I think there are also other modes to explore: I think replacing the file with a IPC pipe could help with logging binary traces piping them directly into cheritrace, also we could limit the set of instructions/events we are interested in (e.g. only instructions that happen to write a capability to memory), as in many cases we don't really need the full instruction trace (e.g. to build the provenance tree).
I think we have the flexibility to handle these things with different trace formats in the backend.

qwattash committed:
Merge typedefs with struct definitions.

qwattash committed:
Fix byte-order conversion for cvtrace format.

qwattash committed:
Cleanup comments.

Remove stale TODOs.
Update comments.

qwattash committed:
Use a local constant variable to avoid repeating qemu_log_instr_enabled in translator loop.

arichardson commented on configure:
```diff
@@ -393,7 +393,7 @@ rdma=""
 gprof="no"
 debug_tcg="no"
-mips_log_instr="yes"
+cheri_log_instr="yes"
```
```suggestion
tcg_log_instr="yes"
```

arichardson commented on configure:
```diff
@@ -1145,9 +1145,9 @@ for opt do
   --disable-debug-tcg) debug_tcg="no"
   ;;
-  --enable-mips-log-instr) mips_log_instr="yes"
+  --enable-cheri-log-instr) cheri_log_instr="yes"
```
```suggestion
  --enable-tcg-log-instr) tcg_log_instr="yes"
```

arichardson commented on configure:
```diff
@@ -1145,9 +1145,9 @@ for opt do
+  --enable-cheri-log-instr) cheri_log_instr="yes"
   ;;
-  --disable-mips-log-instr) mips_log_instr="no"
+  --disable-cheri-log-instr) cheri_log_instr="no"
```
```suggestion
  --disable-tcg-log-instr) tcg_log_instr="no"
```

arichardson commented on configure:
```diff
@@ -6868,8 +6868,8 @@ fi
 fi
-if test "$mips_log_instr" = "yes" ; then
-  echo "CONFIG_MIPS_LOG_INSTR=y" >> $config_host_mak
+if test "$cheri_log_instr" = "yes" ; then
```
```suggestion
if test "$tcg_log_instr" = "yes" ; then
```

arichardson commented on include/exec/helper-gen.h:
```diff
@@ -83,6 +83,10 @@ static inline void glue(gen_helper_, name)(dh_retvar_decl(ret)          \
 #include "tcg-runtime.h"
 #include "plugin-helpers.h"
+#ifdef CONFIG_TCG_LOG_INSTR
+// TODO(am2419): maybe move to tcg or merge with tcg-runtime?
```
Yeah, I think tcg-runtime.h should be fine?

arichardson commented on include/exec/log_instr.h:
```diff
@@ -0,0 +1,255 @@
+ * Log changed capability register.
+ */
+void qemu_log_instr_cap(CPUArchState *env, const char *reg_name,
+                         const cap_register_t *cr);
```
Indentation seems wrong.

arichardson commented on include/exec/log_instr.h:
```diff
@@ -0,0 +1,255 @@
+ * Log changed capability register with integer value.
+ */
+void qemu_log_instr_cap_int(CPUArchState *env, const char *reg_name,
+                             target_ulong value);
```
same

arichardson commented on include/exec/log_instr.h:
```diff
@@ -0,0 +1,255 @@
+ * some trickery to reclaim those.
+ */
+void qemu_log_instr_evt(CPUArchState *env, uint16_t fn, target_ulong arg0,
+                         target_ulong arg1, target_ulong arg2,
```
indentation?

arichardson commented on include/exec/log_instr_helpers.h:
```diff
@@ -0,0 +1,56 @@
+ * logging support is enabled.
+ */
+
+// TODO(am2419): should be TCG_CALL_NO_RWG?
```
I guess NO_WG is safe, not sure about the read.

qwattash replied:
I currently edited the patch to add NO_WG. I read the documentation in tcg/README but it's not entirely clear whether env fields count as globals. My take is that the `cpu_env` register is a global, and TCG will make sure to update it before calling the helper. The fact that we touch fields within `CPUState` should not concern TCG however. The only shady area is whether we risk missing updates to `cpu_env` if we specify `NO_RWG`.

arichardson commented on target/mips/cp0_helper.c:
```diff
@@ -1486,21 +1588,25 @@ void helper_mtc0_taglo(CPUMIPSState *env, target_ulong arg1)
 void helper_mtc0_taghi(CPUMIPSState *env, target_ulong arg1)
 {
     env->CP0_TagHi = arg1; /* XXX */
+    log_instr_cop0_update(env, CP0_REGISTER_29, 1, arg1);
```
Having to add all these logs manually is not ideal, but I guess there isn't really any other solution that results in a smaller diff.

qwattash replied:
I couldn't think of a better way, since they are modified in various different helpers.
The alternative is to revert back to the previous way of scanning all the registers looking for changes. In my opinion that approach is worse as we lose the intent of the current instruction and we could not detect updates that rewrite the same value to the register, this helps having consistent traces without random "holes". Hopefully those helpers do not change a lot and cause many conflicts?

arichardson replied:
I doubt there will be many future changes to the MIPS target so this should be perfectly fine.

arichardson replied:
I doubt there will be many future changes to the MIPS target so this should be perfectly fine.

arichardson commented on target/mips/helper.h:
```diff
@@ -185,13 +185,14 @@ DEF_HELPER_2(mtc0_rtc64, void, env, i64)
-DEF_HELPER_2(instr_stop_user_mode_only, void, env, i64)
-DEF_HELPER_2(instr_stop, void, env, i64)
+#ifdef CONFIG_TCG_LOG_INSTR
+/* Target dependant-helpers */
```
```suggestion
/* Target dependent-helpers */
```

arichardson reviewed (approved):
I'm happy to merge this now if it boots and runs CheriBSD successfully 👍 . Any future improvements logging can be made after this PR has been merged (reviewing changes on top of this huge diff is not ideal).

Huge improvement over the current logging code 🥇

qwattash committed:
Move log instruction info to generic CPUState (take 2).

The per-cpu logging state is moved back to the generic CPUState. A pointer is used to
reference an opaque instruction info. This also makes it easier to implement a ring buffer
for instruction info entries in the future.

Rename the instruction info and per-cpu log state structures.

qwattash committed:
Minor fixes to instruction logging in the TCG translator.

Fix comments regarding TCG flushing.
Add missing unlikely().

qwattash committed:
Update configure

Co-authored-by: Alexander Richardson <Alexander.Richardson@cl.cam.ac.uk>

qwattash committed:
Move instruction logging helpers to tcg-runtime header.

Remove log_instr_helpers header.
Fix instruction logging helpers TCG flags.

qwattash committed:
Fix indentation in log_instr.h

qwattash committed:
Remove cpu_log_instr.h header.

qwattash committed:
Update instruction logging implementetion description comment.

qwattash committed:
Fix logging of the MIPS ASID.

qwattash committed:
Update license headers.

arichardson commented:
@qwattash FYI

Before:
```
 Performance counter stats for './qemu-system-cheri128.before_logging_changes -M malta -kernel /local/scratch/alr48/cheri/output/kernel-mips-hybrid128.CHERI128_MALTA64_MFS_ROOT -m 2048 -nographic -append cheribsd.kernel_startup_benchmark=1' (10 runs):

        819.132249      task-clock (msec)         #    0.475 CPUs utilized            ( +-  0.25% )
               339      context-switches          #    0.413 K/sec                    ( +-  1.11% )
                 3      cpu-migrations            #    0.004 K/sec                    ( +- 15.71% )
             6,996      page-faults               #    0.009 M/sec                    ( +-  0.01% )
     3,314,483,919      cycles                    #    4.046 GHz                      ( +-  0.14% )
     8,634,309,336      instructions              #    2.61  insn per cycle           ( +-  0.13% )
     1,396,245,882      branches                  # 1704.543 M/sec                    ( +-  0.14% )
         7,642,848      branch-misses             #    0.55% of all branches          ( +-  0.15% )

       1.724278541 seconds time elapsed                                          ( +-  0.05% )
```
After:
```
 Performance counter stats for './qemu-system-cheri128.with_logging_changes -M malta -kernel /local/scratch/alr48/cheri/output/kernel-mips-hybrid128.CHERI128_MALTA64_MFS_ROOT -m 2048 -nographic -append cheribsd.kernel_startup_benchmark=1' (10 runs):

        836.020906      task-clock (msec)         #    0.481 CPUs utilized            ( +-  0.14% )
               342      context-switches          #    0.409 K/sec                    ( +-  1.02% )
                 5      cpu-migrations            #    0.006 K/sec                    ( +- 13.22% )
             8,168      page-faults               #    0.010 M/sec                    ( +-  0.01% )
     3,379,754,508      cycles                    #    4.043 GHz                      ( +-  0.13% )
     8,712,187,237      instructions              #    2.58  insn per cycle           ( +-  0.09% )
     1,415,637,261      branches                  # 1693.304 M/sec                    ( +-  0.10% )
         7,803,329      branch-misses             #    0.55% of all branches          ( +-  0.18% )

       1.737390039 seconds time elapsed                                          ( +-  0.06% )
```
